run_unittests.py

```diff
@@ -4446,6 +4476,273 @@ def test_subdir(self):
+        return batfile
+
+    def helper_for_compiler(self, lang, cb):
+        """Helper for generating tests for overriding compilers for langaugages
+        with more than one implementation, such as C, C++, ObjC, ObjC++, and D.
+        """
+        env = get_fake_env('', '', '')
+        getter = getattr(env, 'detect_{}_compiler'.format(lang))
+        if lang not in ['cs']:
+            getter = functools.partial(getter, False)
+        cc = getter()
+        binary, newid = cb(cc)
+        env.config_info.binaries = {lang: binary}
+        compiler = getter()
+        self.assertEqual(compiler.id, newid)
+
+    def test_multiple_native_files_override(self):
+        wrapper = self.helper_create_binary_wrapper('bash', version='foo')
+        config = self.helper_create_native_file({'binaries': {'bash': wrapper}})
+        wrapper = self.helper_create_binary_wrapper('bash', version='12345')
+        config2 = self.helper_create_native_file({'binaries': {'bash': wrapper}})
+        self.init(self.testcase, extra_args=[
+            '--native-file', config, '--native-file', config2,
+            '-Dcase=find_program'])
+
+    def test_multiple_native_files(self):
+        wrapper = self.helper_create_binary_wrapper('bash', version='12345')
+        config = self.helper_create_native_file({'binaries': {'bash': wrapper}})
+        wrapper = self.helper_create_binary_wrapper('python')
+        config2 = self.helper_create_native_file({'binaries': {'python': wrapper}})
+        self.init(self.testcase, extra_args=[
+            '--native-file', config, '--native-file', config2,
+            '-Dcase=find_program'])
+
+    def _simple_test(self, case, binary):
+        wrapper = self.helper_create_binary_wrapper(binary, version='12345')
+        config = self.helper_create_native_file({'binaries': {binary: wrapper}})
+        self.init(self.testcase, extra_args=['--native-file', config, '-Dcase={}'.format(case)])
+
+    def test_find_program(self):
+        self._simple_test('find_program', 'bash')
+
+    def test_config_tool_dep(self):
+        # Do the skip at this level to avoid screwing up the cache
+        if not shutil.which('llvm-config'):
+            raise unittest.SkipTest('No llvm-installed, cannot test')
+        self._simple_test('config_dep', 'llvm-config')
+
+    def test_python3_module(self):
+        self._simple_test('python3', 'python3')
+
+    def test_python_module(self):
+        if is_windows():
+            # Bat adds extra crap to stdout, so the version check logic in the
+            # python module breaks. This is fine on other OSes because they
+            # don't need the extra indirection.
+            raise unittest.SkipTest('bat indirection breaks internal sanity checks.')
+        self._simple_test('python', 'python')
+
+    @unittest.skipIf(is_windows(), 'Setting up multiple compilers on windows is hard')
+    @skip_if_env_value('CC')
+    def test_c_compiler(self):
```

Comment:
@dcbaker Am I understanding correctly that this and other compiler native file tests are skipped completely on the CI because the CI always sets CC/CXX/etc on non-Windows machines? Why don't these tests just unset the env var temporarily with a context manager or something?

I ask because the tests are actually broken on macOS because of this:

```
======================================================================
FAIL: test_cpp_compiler (__main__.NativeFileTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./run_unittests.py", line 175, in wrapped
    return func(*args, **kwargs)
  File "./run_unittests.py", line 5267, in test_cpp_compiler
    self.helper_for_compiler('cpp', cb)
  File "./run_unittests.py", line 5195, in helper_for_compiler
    self.assertEqual(compiler.id, newid)
AssertionError: 'clang' != 'gcc'
- clang
+ gcc
```

(`gcc` on macOS is `clang` in disguise)

Reply:
We certainly could unset them, that just feels like the test doing a lot of bending to try to unset your environment, which you may have set for very important reasons say because cc on your system is suncc or some other unsupported compiler.

Reply:
That seems like a good compromise. I'm trying to remember, but the specific problems I ran into might have been related to ICC